File: python/algorithms/CountingSort.py
def countSort(arr): 
  
    output = [0 for i in range(500)] 
    count = [0 for i in range(500)] 
  
    ans = ["" for _ in arr] 
    for i in arr: 
        count[ord(i)] += 1
  
    for i in range(256): 
        count[i] += count[i-1] 
    for i in range(len(arr)): 
        output[count[ord(arr[i])]-1] = arr[i] 
        count[ord(arr[i])] -= 1
  
    for i in range(len(arr)): 
        ans[i] = output[i] 
    return ans  
  
arr = "thisisawesome"
ans = countSort(arr) 
print "Sorted character array is %s"  %("".join(ans)) 
 

File: java/algorithms/CountingSort.java
/**
 * Date: 2018-10-01
 */

public class CountingSort {
	public static char[] sort(char[] input, boolean replaceInput) 
	{
		char[] sorted = sort(input);
		
		if( replaceInput ) {
			// Copy the output array to input array
	        for (int i = 0; i < sorted.length; i++) 
	            input[i] = sorted[i]; 
		}
		
		return sorted;
	}
	
	public static char[] sort(char[] input) 
    { 
	    // Create output array with size of input 
        char output[] = new char[input.length]; 
  
        // Create array to hold occurences of each
        // character (char = 256)
        int countForChar[] = new int[256]; 
        
        // Initialize to 0 (java defaults to zero but better to be safe)
        for (int i=0; i<256; ++i) 
        	countForChar[i] = 0; 
  
        // Loop trough all characters in inputstore count of each character 
        for (int i = 0; i < input.length; i++) 
            countForChar[input[i]]++; 
  
        int outputPos = 0;
        for( int i = 1; i <= 255; i++ ) {
        	if( countForChar[i] > 0 ) {
        		for( int j = 0; j < countForChar[i]; j++) {
        			output[outputPos] = (char)i;
        			outputPos++;
        		}
        	}
        }
        
        return output;
    } 
}
